pysat/tests/instrument_test_class.py

```diff
@@ -47,18 +49,95 @@ def initialize_test_inst_and_date(inst_dict):
 
+    def setup_class(self):
+        """Initialize the testing setup once before all tests are run."""
+
+        # Use a temporary directory so that the user's setup is not altered.
+        self.tempdir = tempfile.TemporaryDirectory()
+        self.saved_path = pysat.params['data_dirs']
+        pysat.params['data_dirs'] = self.tempdir.name
+        return
+
+    def teardown_class(self):
+        """Clean up downloaded files and parameters from tests."""
+
+        pysat.params['data_dirs'] = self.saved_path
+        self.tempdir.cleanup()
+        del self.saved_path, self.tempdir
+        return
+
+    def apply_marks_to_tests(self, inst_loc, user_info=None):
+        """Apply custom instrument lists to each test method as a pytest mark.
+
+        Parameters
+        ----------
+        inst_loc : python subpackage
+            The location of the instrument subpackage to test, e.g.,
+            'pysat.instruments'
+        user_info : dict or NoneType
+            Nested dictionary with user and password info for instrument module
+            name.  If None, no user or password is assumed. (default=None)
+            EX: user_info = {'jro_isr': {'user': 'myname', 'password': 'email'}}
+
+        """
+
+        # Attach location of package to test object for later reference.
+        self.inst_loc = inst_loc
+
+        # Find all instruments for testing from user-specified location.
+        instruments = generate_instrument_list(inst_loc=inst_loc,
+                                               user_info=user_info)
+
+        # Find all methods in the standard test class.
+        method_list = [func for func in dir(self)
+                       if callable(getattr(self, func))]
+        # Search tests for iteration via pytestmark, update w/ instrument list.
+        for method in method_list:
+            if hasattr(getattr(self, method), 'pytestmark'):
+                # Get list of names of pytestmarks.
+                n_args = len(getattr(self, method).pytestmark)
+                mark_names = [getattr(self, method).pytestmark[j].name
+                              for j in range(0, n_args)]
+                # Add instruments from your library.
+                if 'all_inst' in mark_names:
+                    mark = pytest.mark.parametrize("inst_name",
+                                                   instruments['names'])
+                    getattr(self, method).pytestmark.append(mark)
+                elif 'download' in mark_names:
+                    mark = pytest.mark.parametrize("inst_dict",
+                                                   instruments['download'])
+                    getattr(self, method).pytestmark.append(mark)
+                elif 'no_download' in mark_names:
+                    mark = pytest.mark.parametrize("inst_dict",
+                                                   instruments['no_download'])
+                    getattr(self, method).pytestmark.append(mark)
+        return
```

Comment:
```suggestion
        return instruments
```
We could return the `instruments` object if needed so that it doesn't need to be run separately when needed for additional tests.  This would clean up the code, but may be more confusing when writing the docs.  Thoughts?

Reply:
Using the return statement.